can: add a CAN stack for RIOT

vincent-d opened the pull request:
Hi,

This PR propose a CAN stack for RIOT.

It includes a device driver interface (drivers/include/can/candev.h) and a reference driver for native cpu and board (under Linux only using socketCAN).

The architecture of the device driver is inspired by netdev2, with a thread per device communicating with messages with the upper layer threads.

A basic conn API is also provided for user applications, for raw CAN and isotp.

This PR implements only the data link layer (raw CAN) and provide support for simultaneous reception from multiple user threads.

A common area is used to allocate CAN packets (frame + stack data) (implementation in sys/can/pkt.c) and the routing of the messages to the user threads is done by the "router" part (sys/can/router.c).

We are still working on 2 others drivers for stm32 internal CAN controller and MCP2515 (SPI CAN controller) and on a common interface to manage CAN transceivers.

Still to do:
- add unit tests (ongoing)
- ~~power management and wake up from the CAN bus~~
- ~~improve bittiming setting~~
- ~~implementation of ISO-TP~~

---

You can test under Linux by setting up virtual CAN interface:

```
sudo modprobe vcan
sudo ip link add vcan0 type vcan
sudo ip link add vcan1 type vcan
sudo ip link set vcan0 up
sudo ip link set vcan1 up
```

Then you can run one the test app (tests/conn_cann) as usual:

```
make all
make term
```

Sending on can0 from RIOT, frame id 0x100, data 0x11 0x22:

```
can send 0 0x100 0x11 0x22
```

Receiving on can0 from RIOT, with thread 0, frame id 0x101:

```
can recv 0 0 0x101
```

Sending from Linux (with canutils), frame id 0x100, data 0x11 0x22:

```
cansend vcan0 100#1122
```

Receiving from Linux (any frame)

```
candump vcan0
```

---

Comments are welcommed :)

toonst commented:
Great to see it's here!

miri64 commented:
Since googling for CAN just gives me a bunch of pictures of beverage containers I would appreciate some information about this stack ;-)

kYc0o commented:
Try this [one](https://en.m.wikipedia.org/wiki/CAN_bus) ;)

miri64 commented:
thanks :-)

kaspar030 commented on drivers/include/can/candev.h:
```diff
@@ -0,0 +1,181 @@
+    const struct candev_driver *driver;    /**< ptr to that driver's interface. */
+    candev_event_cb_t event_callback;      /**< callback for device events */
+    void *isr_arg;                         /**< argument to pass on isr event */
+    kernel_pid_t pid;                      /**< device thread pid */
```
is this necessary?

netdev2 has been designed to not force usage of a thread per device. maybe that architecture can be re-used here?

vincent-d replied:
if we want the candev to be completely thread-agnostic, we can remove that from this struct, then we would need to add it into a struct in the can/device.h file. 

I'll propose something.

kaspar030 commented on sys/Makefile.include:
```diff
@@ -69,4 +69,8 @@ ifneq (,$(filter printf_float,$(USEMODULE)))
 endif
 
+ifneq (,$(filter can,$(USEMODULE)))
+    USEMODULE_INCLUDES += $(RIOTBASE)/sys/include/can/
```
below, all includes are ```can/file.h```. Doesn't that make this include path redundant?

vincent-d replied:
addressed

kaspar030 commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+#include "xtimer.h"
+#endif
+
+#define ENABLE_DEBUG (1)
```
0 for merging

vincent-d replied:
addressed

lebrush commented:
Awesome!! I'll have a look at the code whenever I've a chance. 

In case you are still working on a stm32 driver, I've a working version I wrote time ago for a previous RIOT version. Have a look [here](https://github.com/lebrush/RIOT/commit/5995512ed3b724874873947c02ff65e05118d4a0) for the driver and [here](https://github.com/lebrush/RIOT/commit/2e310c9e7957625c549a034db0e9404ddae4fa78) for the necessary board defines configuration.

miri64 commented:
Apart from `get_filter`, `set_filter`, and `abort` (and the `filter` member) the driver struct doesn't look too dissimilar to `netdev2_driver_t`. Can we model the three functions as options (NETOPT_FILTER and NETOPT_ABORT)? I'm unsure about the `filter` member, but putting it into the function struct (which ideally would be `const` so we can benefit from compiler optimizations) seems odd to me anyway. So can we move it to the `candev_t` struct? Then it would be possible to model them just as `netdev2`s.

vincent-d commented:
A few words about CAN and filters.

With CAN you don't address nodes but frames, then every frame is broadcasted in the bus. Every CAN controller provide hardware filtering, to allow a node to receive only the frames it wants to receive.

The `nb_filter` member was here to set the number of hardware filters available in the controller. It's not used at the moment, and I'm not sure anymore if it makes sense anyway.

`set_filter` and `remove_filter` are then used to set and remove a hardware filter in the controller. The reason why we chose not using the generic `set` is that the `set_filter` is supposed to return an id associated to the filer we could use later on when receiving a frame. Knowing which filter was triggered might help to route the frame to the right user even if it's not used at the moment. Then `remove_filter` was build symetrically to remove a filter, and otherwise we would need an option like NETOPT_REMOVE_FILTER.

Another difference between `candev_driver_t` and `netdev2_driver_t` is that we remove the `recv` function. Frame reception is then achieved through the `event`. The received frame is passed as parameter. 

Which leads to another difference, there is an additional parameter in the `event` function, which is used when sending, to return the handle of the sent frame and then be able to know wich frame has been sent correctly (event CANDEV_EVENT_TX_CONFIRMATION) or has timeouted (event CANDEV_EVENT_TIMEOUT_TX_CONF), or when receiving, to return the received frame as I previously wrote.

I'm wondering whether merging `candev_driver_t` and `netdev2_drivert_t` would actually make sense. I'm pretty convinced that this CAN stack could share some code with the network stack, but on the low level, a CAN controller will never be able to plugged to an IP network stack, and and radio module will never be connected to a CAN bus, so which benefit could we find in merging the both driver api ? I'm not against the idea, but I have trouble seeing the advantages.

miri64 commented:
Okay. Thanks for the explanation :)

lebrush commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,673 @@
+    candev_linux_t *dev = (candev_linux_t *)candev;
+
+    mutex_init(&dev->rx_mutex);
+    mutex_init(&dev->rx_mutex);
```
the mutex was already initialized in the line above

vincent-d replied:
both mutex are actually useless, I will remove them

lebrush commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,87 @@
+    struct can_bittiming *timing;
+    mutex_t tx_mutex;
+    mutex_t rx_mutex;
+    canid_t can_id_sent;
```
this is never used. What is it planned for?

vincent-d replied:
will be removed

lebrush commented on drivers/include/can/candev.h:
```diff
@@ -0,0 +1,180 @@
+ * @param[in] type          type of the event
+ * @param[in] arg           event argument
+ */
+typedef void (*candev_event_cb_t)(candev_t *dev, candev_event_t event, void *arg);
```
missing doc for `dev`

lebrush commented on drivers/include/can/candev.h:
```diff
@@ -0,0 +1,180 @@
+ */
+typedef struct candev_driver {
+    /**
+     * The nnumber of HW filters available
```
doc: `nnumber`

lebrush commented on drivers/include/can/candev.h:
```diff
@@ -0,0 +1,180 @@
+     * @brief Send packet
+     *
+     * @param[in] dev       CAN device descriptor
+     * @param[in] packet    CAN packet to send
```
parameter is named `pkt` in the declaration (also below)

lebrush commented on sys/auto_init/can/auto_init_candev.c:
```diff
@@ -0,0 +1,94 @@
+#ifdef MODULE_CAN_LINUX
+/* 500kbps */
+static const struct can_bittiming timing_linux = {
+#ifdef MODULE_CAN_LINUX
```
this is redundant with the upper check (line 34)

lebrush commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+    kernel_pid_t my_pid = thread_getpid();
+
+    /* initialize conn structure */
+    conn->pid = my_pid;
```
why not assigning directly from `thread_getpid()`? 'my_pid` is not used anymore within the method.

miri64 replied:
Or `sched_active_pid` directly ;-).

lebrush commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+
+    kernel_pid_t my_pid = conn->pid;
+    int ifnum = conn->ifnum;
+    can_family_t family = conn->family;
```
I think you don't need this three auxiliary variables

lebrush commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+    if (timeout > 0) {
+        timeout_msg.type = _TIMEOUT_MSG_TYPE;
+        timeout_msg.content.value = _TIMEOUT_MSG_VALUE;
+        xtimer_set_msg(&timeout_timer, timeout, &timeout_msg, conn->pid);
```
also on the direction of the question of @kaspar030. If `conn->pid` is assigned to the process which initializes the connection, do you expect a different process to call `conn_can_recv`? If yes, why should then the connection creator be alerted of a timeout? (maybe I'm not far enough in the code yet...)

vincent-d replied:
Yes, the idea was one conn == one thread.

lebrush replied:
maybe I missed something, please correct me if I'm wrong: why do you need to store the PID of the creator process in `conn->pid` if it's only used for method which are called in the same process? what do I miss?

vincent-d replied:
good point, but I wonder if it could make sense to allow creating/closing the connection from another thread than the one which receive. Then we could just remove the pid from the conn and simplify the code.

lebrush commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+    switch (family) {
+#ifdef MODULE_CAN_RAW
+        case CONN_CAN_RAW:
+            if (max_len != sizeof(struct can_frame)) {
```
should this not be `max_len < sizeof(struct can_frame)`?

lebrush commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+                    xtimer_remove(&timeout_timer);
+                    if (msg.content.value == _TIMEOUT_MSG_VALUE) {
+                        return -ETIMEDOUT;
+                    }
```
is it fine, that there's no `break` here?

vincent-d replied:
Yes, the message will then be requeued

lebrush commented on sys/can/conn_can.c:
```diff
@@ -0,0 +1,331 @@
+    kernel_pid_t my_pid = conn->pid;
+    int ifnum = conn->ifnum;
+    can_family_t family = conn->family;
+
```
same as [comment above](https://github.com/RIOT-OS/RIOT/pull/5793/files#r77335613)

lebrush commented on sys/can/dll.c:
```diff
@@ -0,0 +1,274 @@
+
+    msg.type = CAN_MSG_SET;
+    msg.content.ptr = (void *)opt;
+    msg_send_receive(&msg, &reply, opt->context);
```
I've a question regarding implementation here. 

The flow I've observed is: you call this method, it sends a message to `_can_device_thread`, it sets the driver option, then returns a message to this process, this process returns the return value of `dev->driver->set(...)`. Why not calling it here directly? (same for similar functions)

vincent-d replied:
This is a copy of the architecture of gnrc_netdev2. And here the dll keep track of ifnum and pid, so it doesn't have a pointer to the device, then every operation from upper layer goes through the `_can_device_thread`.

lebrush commented:
I've some questions regarding implementation and i'm still thinking about the similarities with `netdev2` but I think that this looks really promising and is a great addition to RIOT :-)

vincent-d commented:
@lebrush  Sorry we forgot to clen up some old stuff in the candev_linux, should be better now. Thanks for your remarks

vincent-d commented:
> I've some questions regarding implementation and i'm still thinking about the similarities with netdev2 

Do not hesitate to ask. And I hope that you guys can help pointing similarities and which code could be shared with other parts of RIOT.

> I think that this looks really promising and is a great addition to RIOT :-)

Thank you :-)

vincent-d commented:
pid removed from struct conn_can.

I also fixed unsetting filters and updated the conn_can test app to be able to test with multiple receive threads.

The candev driver for native has been improved too, it now uses socketCAN library (installation procedure available in tests/conn_can/README.md).

LudwigKnuepfer commented:
Regarding the device configuration in native:
Up to now it was customary to pass such paths as a command line parameter/option.
As far as I see there is no particular reason to break this pattern here.
Therefore I would kindly ask you to adopt this pattern for this driver as well or argue against it ;)

vincent-d commented:
@LudwigKnuepfer I see at least one issue. We use CAN_DLL_NUMOF constant to initialize the device threads, so we must known at compile time how much interface there will be. And it's important for us to be able to test with multiple devices. Then passing interface names as parameters would mean to rework completely the auto init and config parts.

LudwigKnuepfer commented:
@vincent-d I don't think you would need to rework anything completely.. From my perspective it's OK to set the number of devices at compile time. From a quick glance at the code it should not be problem to make the names dynamic then.

LudwigKnuepfer commented:
BTW: Style-wise I'd consider replacing the `CAN_DLL_NUMOF` macro with a `const` variable an improvement.

vincent-d commented:
@LudwigKnuepfer OK, if we keep the number of interface at compile time, then we can do the change.

For the macro vs const, I had the impression that macro style was widely used in RIOT for this kind of constant, am I wrong ?

LudwigKnuepfer commented:
> For the macro vs const, I had the impression that macro style was widely used in RIOT for this kind of constant, am I wrong ?

I think you are probably right, but that does not mean it's a good thing ;)
But this is not the place to discuss policies - I just shared my personal view and you are free to keep it as it is.

vincent-d commented:
Lot of refactoring and rework.

ISOTP implementation added. 

I still need to work on testing it a bit more!

vincent-d commented:
Rebased on last master

emmanuelsearch commented:
Does anyone out there (except @vincent-d ) have hardware available to test this?

vincent-d commented:
@emmanuelsearch This PR provide only a native driver and the CAN stack. I have two others PR in my tree for STM32 CAN periph and MCP2515 that I'll open soon.

vincent-d commented:
I have added the support for transceiver into the CAN stack, with drivers for tja1042 (HS CAN transceiver) and NCV7356 (SW CAN transceiver).

Power management have been improved as well, with the transceiver support.

LudwigKnuepfer commented:
I have a couple of MCP2515 breakout boards lying around - if anyone wants to borrow them for testing feel free to contact me.

vincent-d commented:
Power management added. After some inactivity time (which can be set up per device), the device is powered off and the transceiver set in standby.

OlegHahm commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,103 @@
+extern "C" {
+#endif
+
+#if defined(__linux__) //SocketCAN is supported only on Linux
```
Please use C style comments and add documentation to all header files.

OlegHahm replied:
(The macros below are missing doc.)

OlegHahm commented on drivers/include/can/candev.h:
```diff
@@ -0,0 +1,207 @@
+ *          upper layer
+ */
+typedef enum {
+    CANDEV_EVENT_NOEVENT,
```
Missing doc.

OlegHahm commented on drivers/ncv7356/ncv7356_trx.c:
```diff
@@ -0,0 +1,80 @@
+        gpio_set(ncv7356->mode1_pin);
+        ret = 0;
+        break;
+    */
```
I highly discourage to commit any dead code - but at least it should be documented why this block is disabled.

vincent-d replied:
addressed

OlegHahm commented on sys/include/can/can.h:
```diff
@@ -0,0 +1,128 @@
+
+#else
+
+#define CAN_MAX_DLEN 8
```
I know I'm a bit manic in this regard, but I always prefer to have macro definitions with parens (or at least consistent throughout a module.)

OlegHahm commented on sys/auto_init/can/auto_init_can.c:
```diff
@@ -0,0 +1,101 @@
+        case CAN_DRIVER_STM32:
+            break;
+        case CAN_DRIVER_MCP2515:
+            break;
```
Wouldn't it be cleaner to only call a generic CAN init function here and put the device specific init code into separate files (like we do for netdev)? Otherwise this code here is prone to become a "IFDEF hell" 👿  very soon.

vincent-d replied:
There is something I don't like with the way we did this auto init. The main reason why we put eveything into the same function was to be able to control the order of the interfaces. 
But then we have this big `can_iface_t candev_ifnum[]` array in the can_conf.h file which is included from a lot of places, just to have the CAN_DLL_NUMOF variable defined.

I will think about a better way to initialize everything.

OlegHahm replied:
I completely agree that auto_init itself is suboptimal for now.

OlegHahm commented on sys/can/conn/raw.c:
```diff
@@ -0,0 +1,280 @@
+        }
+    }
+    else {
+#ifdef MODULE_XTIMER
```
Just curious: do you have any specific use case in mind where you want to disable xtimer?

vincent-d replied:
No, and now that you say it, I think it would be safer to always enable it.

OlegHahm commented on sys/can/dll.c:
```diff
@@ -0,0 +1,460 @@
+ *
+ *
+ * @file
+ * @brief       CAN DLL
```
Well, this is very brief.

OlegHahm commented on sys/can/device.c:
```diff
@@ -0,0 +1,598 @@
+    }
+    else {
+        if (timing->bitrate > 800000) {
+            spt = 750;
```
What are these magic numbers?

vincent-d replied:
These are the sample point value recommended by the CiA (CiA 301). But I just discovered this was a bit outdated, last document recommends 87.5% for every baudrate up to 1Mbit/s.

I will fix that and add some comments in the code to explain.

OlegHahm commented on sys/can/device.c:
```diff
@@ -0,0 +1,598 @@
+    uint32_t rate_error;
+    uint32_t min_rate_error;
+
+    assert(timing != NULL && timing->bitrate != 0);
```
My parens-mania again...

OlegHahm commented on sys/can/dll.c:
```diff
@@ -0,0 +1,460 @@
+
+    if (filter == NULL) {
+        return -ENOENT;
+    }
```
Is it really necessary to distinguish different error codes? Is it realistic that the system handles these errors differently at runtime?

vincent-d replied:
I don't have any use case in mind where it could be really usefull to distinguish error codes, except for debugging. I'm also wondering if I should not replace all theses tests by simple asserts.

OlegHahm replied:
Using asserts seems to be reasonable to me.

OlegHahm commented on sys/can/pkt.c:
```diff
@@ -0,0 +1,287 @@
+ *
+ *
+ * @file
+ * @brief       Definitions high-level CAN interface
```
I think this line is inaccurate.

OlegHahm commented on sys/can/pkt.c:
```diff
@@ -0,0 +1,287 @@
+/*
```
Hm, I don't have a concrete proposal right now, but I really dislike the code duplication with pktbuf here. Can you imagine a way to consolidate some common code?

vincent-d replied:
I agree, I would love to reuse the pktbuf from gnrc, but I still can't find an easy way to do it.

OlegHahm replied:
Maybe you can recap your thought about this problem so far at the next virtual developer meeting. What do you think? I think it's worth a more general discussion since other developers may need a similar data structure.

vincent-d replied:
Sure. For the record I will recap here why I need a memory pool for dynamic allocation.

First we need to allocate memory for incoming and outcoming CAN frames. This was the very first usage of the pkt module, and the first implementation was an array of frames. CAN frames are composed of at most 8 data bytes plus a CAN id.

Then I needed to allocate some memory for filter lists. They are used to make a link between incoming frames and user code, which can subscribe to some frames. Filters are also quite small, a filter is just a struct containing two `uint32_t`, with what needed around to route the frame to the right user.

The last usage is for isotp buffers. ISO-TP (transport protocol) is used to send more than 8 bytes on CAN. Up to 4095 bytes can theoritically be sent through an isotp channel. For this last use case, the allocated buffers can be bigger than in the two previous ones. Even if in the use cases I known it rarely goes above 500 bytes.

OlegHahm replied:
Thanks!

vincent-d commented:
@OlegHahm Thanks for the first review, I will address the comments as soon as possible.

OlegHahm commented:
@haukepetersen, maybe we can discuss this next week?

vincent-d commented:
I reworked the init process. It's now closer to the saul init process. Each device needs to be named and configured in a *_params.h file which is included in the auto_init_*.c file.

The drawback of this approach is that the interface number is not static anymore, the user code must then retrieve the interface number from the device name. But it makes the autoinit code much more clean in my opinion.

vincent-d commented:
If nobody object, I'll squash. The commit list became a bit big.

kYc0o commented:
Yes squash please. However, if one of the commits is for addressing a comment, leave it if possible.

vincent-d commented:
Rebased.

- Minor fixes with conn_can_isotp
- Improve CAN_DLL_NUMOF definition during build
- fix atomic calls (-> stdatomic)

vincent-d commented:
- Lot of doxygen fixes
- rework trx enum values

vincent-d commented:
@OlegHahm @haukepetersen the last two commits use `gnrc_pktbuf` as the allocator. It must add some overhead in memory but has the advantage to share the memory management code.

vincent-d commented:
Doxygen reorganized.

Jenkins is not able to find libsocketcan but I don't know how to fix that. 

The rest should be OK.

vincent-d commented:
Rebased.

Does someone want to start reviewing it ?

vincent-d commented:
ping!

Is there someone interested in reviewing that one? How could we help to have it merged?

vincent-d commented:
I squashed as the commit list became really long. I kept separate commits for the memory allocation part using gnrc_pktbuf.

aabadie commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,553 @@
+    struct can_frame rcv_frame;
+    candev_linux_t *dev = (candev_linux_t *)candev;
+
+    //TODO: Poll to detect others errors than data available? (data = POLLIN)
```
C++ style comments + is the todo still valid ?

aabadie commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,553 @@
+    DEBUG("candev_native _isr: CAN SIGIO interrupt received, sock = %i\n", dev->sock);
+    nbytes = real_read(dev->sock, &rcv_frame, sizeof(struct can_frame));
+
+    if (nbytes < 0) {   //SIGIO signal was probably due to an error with the socket
```
c++ style comment

aabadie commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,553 @@
+        DEBUG("candev_native _isr: error frame\n");
+        candev_event_t evt = _can_error_to_can_evt(rcv_frame);
+        if ((evt != CANDEV_EVENT_NOEVENT) && (dev->candev.event_callback)) {
+                dev->candev.event_callback(&dev->candev, evt, NULL);
```
indent is off

aabadie commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,553 @@
+{
+    candev_linux_t *dev = (candev_linux_t *)candev;
+
+
```
extraline that can be removed

aabadie commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,121 @@
+
+#ifndef CANDEV_LINUX_H
+#define CANDEV_LINUX_H
+
```
extra line

aabadie commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * candev base structure
+     */
+    candev_t candev;
```
style: I would prefer doxygen documentation this way for struct attributes:
```c
typedef struct candev_linux {
    candev_t candev; /**< candev base structure */
[..]
}
```

aabadie commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,121 @@
+#endif
+
+/**
+ * The candev_linux struct
```
use `@brief`

aabadie commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,121 @@
+/**
+ * @brief Device specific initialization function
+ *
+ * @param dev   the device to initialize
```
specify if it's input or output parameter (`@param[in]`, `@param[out]`)

aabadie commented on cpu/native/include/candev_linux_params.h:
```diff
@@ -0,0 +1,44 @@
+#endif
+
+/**
+ * Default parameters (device names)
```
`@brief`

aabadie commented on cpu/native/include/candev_linux_params.h:
```diff
@@ -0,0 +1,44 @@
+#endif
+
+#endif /* CANDEV_LINUX_PARAMS_H */
+
```
extra line

aabadie commented on drivers/include/can/trx.h:
```diff
@@ -0,0 +1,109 @@
+ * @author      Aurelien Gonce <aurelien.gonce@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ */
+#ifndef CAN_TRX_H_
```
trailing underscore

aabadie commented on drivers/include/can/trx.h:
```diff
@@ -0,0 +1,109 @@
+}
+#endif
+
+#endif /* CAN_TRX_H_ */
```
underscore

aabadie commented on drivers/include/ncv7356_trx.h:
```diff
@@ -0,0 +1,87 @@
+ * @author      Aurelien Gonce <aurelien.gonce@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ */
+#ifndef NCV7356_TRX_H_
```
underscore, and probably in other places

aabadie commented on sys/can/router.c:
```diff
@@ -0,0 +1,399 @@
+typedef struct filter_el {
+    can_reg_entry_t entry;
+    canid_t can_id;          /**< CAN ID of the element */
+    canid_t mask;
```
missing doc

aabadie commented on sys/include/can/dll.h:
```diff
@@ -0,0 +1,110 @@
+ * @brief Dispatch RX error from a device
+ *
+ * Dispatch RX error from a device to receivers threads
+ * which have subsribed to frames on that interface
```
subscribed

aabadie commented on sys/include/can/doc.txt:
```diff
@@ -0,0 +1,23 @@
+ *
+ * The low-level interface, candev, must be implemented by controler drivers.
+ * The optionnal transceiver support can also be activated. Transceiver drivers must
+ * then implenebt the trx_can interface.
```
implement

aabadie commented on sys/include/can/isotp.h:
```diff
@@ -0,0 +1,197 @@
+/**
+ * @ingroup     can
+ * @defgroup    isotp ISOTP
+ * @brief ISO transport protocol over CAN (ISO15765)
```
please align with others

aabadie commented on sys/include/can/isotp.h:
```diff
@@ -0,0 +1,197 @@
+    uint8_t bs;     /**< block size */
+    uint8_t sn;     /**< current sequence number */
+    int tx_handle;  /**< handle of the last sent frame */
+    gnrc_pktsnip_t *snip;
```
doc missing

aabadie commented on sys/include/can/isotp.h:
```diff
@@ -0,0 +1,197 @@
+/**
+ * @brief Release a bound isotp channel
+ *
+ * It unset the filter on the DLL and remove the channel from the list
```
unset**s**, remove**s**

aabadie commented on sys/include/can/isotp.h:
```diff
@@ -0,0 +1,197 @@
+/**
+ * @brief Bind an isotp channel
+ *
+ * It initialize the channel, set the filter on the DLL and adds the
```
initialize**s**

aabadie replied:
or remove `It` here and below

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+    atomic_uint ref_count;   /**< Reference counter (for rx frames) */
+    int handle;              /**< handle (for tx frames */
+    struct can_frame frame;  /**< CAN Frame */
+    gnrc_pktsnip_t *snip;
```
doc

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,269 @@
+ * @file
+ * @brief       Definitions high-level CAN interface
+ *
+ * This files define the hig-level CAN interface to send and receive RAW CAN frame.
```
file

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,269 @@
+ * on the interface @p ifnum.
+ * The user thread will then receive msg via IPC on reception of frame matching @p filters.
+ *
+ * Currently only single frame ID (i.e. filters->can_mask = 0xFFFFFFFF) are supported.
```
shouldn't it be `is supported`

aabadie replied:
or IDs

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,269 @@
+/**
+ * @brief Send a CAN frame
+ *
+ * This functions sends a CAN @p frame through the @p ifnum interface. The result is
```
Sends a CAN (remove This functions)

aabadie commented on tests/conn_can/main.c:
```diff
@@ -0,0 +1,658 @@
+    }
+
+
+    receive_pid[RCV_THREAD_NUMOF - 1] = sched_active_pid;
```
I don't really understand this. Why not just creating the 2 receiving threads in the previous loop and run the shell in the main thread ?

vincent-d replied:
When I started that app, I had some troubles with the thread priorities. I needed to reduce shell priority. I'm not sure if it applies anymore, though.

vincent-d commented:
Addressed @aabadie comments + minor docs and cosmetics clean up.

aabadie commented on drivers/include/can/trx.h:
```diff
@@ -0,0 +1,108 @@
+typedef struct trx_driver trx_driver_t;
+
+/**
+ * Generic transceiver descriptor
```
`@brief`

aabadie commented on drivers/include/can/trx.h:
```diff
@@ -0,0 +1,108 @@
+} trx_t;
+
+/**
+ * Generic transceiver driver
```
`@brief`

aabadie commented on drivers/include/ncv7356_trx.h:
```diff
@@ -0,0 +1,86 @@
+#endif
+
+/**
+ * ncv7356 trx descriptor
```
`@brief`

aabadie commented on drivers/include/ncv7356_trx.h:
```diff
@@ -0,0 +1,86 @@
+    /**
+     *  set mode interface
+     */
+    trx_t trx;
```
use /**< */

aabadie commented on drivers/include/tja1042_trx.h:
```diff
@@ -0,0 +1,80 @@
+    /**
+     *  set mode interface
+     */
+    trx_t trx;
```
`/**< */`

aabadie commented on sys/include/can/can.h:
```diff
@@ -0,0 +1,149 @@
+};
+
+/**
+ * CAN bit-timing parameters
```
`@brief`

aabadie commented on sys/include/can/can.h:
```diff
@@ -0,0 +1,149 @@
+};
+
+/**
+ * CAN harware-dependent bit-timing constant
```
`@brief`

aabadie commented on sys/include/can/conn/isotp.h:
```diff
@@ -0,0 +1,199 @@
+
+#ifndef CONN_CAN_ISOTP_MBOX_SIZE
+/**
+ * Mailbox size of a conn_can_isotp_t
```
`@brief`

aabadie commented on sys/include/can/conn/isotp.h:
```diff
@@ -0,0 +1,199 @@
+
+#ifndef CONN_CAN_ISOTP_MBOX_SIZE
+/**
+ * Mailbox size of a conn_can_isotp_t
```
`@brief`

aabadie commented on sys/include/can/conn/raw.h:
```diff
@@ -0,0 +1,140 @@
+
+#ifndef CONN_CAN_RAW_MBOX_SIZE
+/**
+ * Mailbox size of a conn_can_raw_t
```
`@brief`

aabadie commented on sys/include/can/conn/raw.h:
```diff
@@ -0,0 +1,140 @@
+#define CONN_CAN_RAW_MBOX_SIZE (16)
+#endif
+
+/* flags values */
```
Can we group those flags using `@name` ?

aabadie commented on sys/include/can/device.h:
```diff
@@ -0,0 +1,115 @@
+#endif
+
+/**
+ * Parameters to initialize a candev
```
`@brief`

aabadie commented on sys/include/can/device.h:
```diff
@@ -0,0 +1,115 @@
+} candev_params_t;
+
+/**
+ * candev descriptor to pass to the device thread
```
`@brief`

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,268 @@
+#endif
+
+/**
+ * Default value for undefined interface number
```
`@brief`

aabadie commented on sys/include/can/can.h:
```diff
@@ -0,0 +1,149 @@
+/** @} */
+
+/**
+ * CAN operational and error states
```
`@brief` for consistency

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+void can_pkt_free(can_pkt_t *pkt);
+
+/**
+ * Allocate a @p can_rx_data_t and initialize it with gieven parameters
```
`@brief`

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+can_rx_data_t *can_pkt_alloc_rx_data(void *data, size_t len, void *arg);
+
+/**
+ * Free rx data previously allocated by can_pkt_alloc_rx_data()
```
`@brief`

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+void can_pkt_free_rx_data(can_rx_data_t *data);
+
+/**
+ * Allocate @p size bytes and return the pointer
```
`@brief`

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+void *can_pkt_buf_alloc(size_t size);
+
+/**
+ * Free the data allocated by can_pkt_buf_alloc()
```
`@brief`

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+void can_pkt_buf_free(void *data, size_t size);
+
+/**
+ * Retrieve the packet corresponding to @p handle
```
`@brief`

aabadie commented on tests/conn_can/README.md:
````diff
@@ -0,0 +1,69 @@
+make
+make flash
+make term
+```
````
it would be great to give more commands showing the usage of the shell (`can`)

aabadie replied:
Thanks for adding the extra commands. Can you also add the commands related to the use of canutils between Linux and native ?

aabadie commented:
The test application works on native.

I was just wondering if it would prefer to split this PR in 3:
* First one with core implementation of the can stack. This one should be at least testable on Linux with RIOT running native
* Second providing the driver for `ncv7356`
* Third providing the driver for `tja1042`

The idea is to be able to merge the first one fast because anyone can test (like I just did), it also has the benefit of reducing the diff => simpler to review.

Is it ok for you @vincent-d ?

vincent-d commented:
@aabadie sounds reasonable! I'll try to update that next week.

vincent-d commented:
@aabadie comments addressed + small bugfixes

I also removed tja1042 and ncv7356 drivers and will open 2 PRs for them.

aabadie commented:
great thanks. That was fast ! I'll have another look asap. By the way, I fear it will be a bit short to have it in the release (feature freeze is today) except if others think it could.

aabadie commented on sys/can/Makefile:
```diff
@@ -0,0 +1,11 @@
+MODULE = can
```
This line is not needed since the module name and the directory are the same

aabadie commented on sys/include/can/doc.txt:
```diff
@@ -0,0 +1,23 @@
+ * a user interface.
+ *
+ * The low-level interface, candev, must be implemented by controler drivers.
+ * The optionnal transceiver support can also be activated. Transceiver drivers must
```
optional

aabadie commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,550 @@
+static void _callback_can_sigio(int sock, void *arg);
+
+candev_linux_conf_t candev_linux_conf[CAN_DLL_NUMOF] = {
+#if CAN_DLL_NUMOF >= 1
```
This is limited to 2 can interfaces. Is it an expected limitation (I don't know much on CAN) ?

I'm wondering if it's possible to define the list of can interface outside (similar to drivers params structs):
```c
#define CAN_DLL_INTERFACES             \
    {                                  \
            .interface_name = "vcan0", \
    },                                 \
    {                                  \
            .interface_name = "vcan1", \
    }                                  \

candev_linux_conf_t candev_linux_conf[] = {
    CAN_DLL_INTERFACES
}

#define CAN_DLL_NUMOF    (sizeof(candev_linux_conf) / sizeof(candev_linux_conf[0]))
```

aabadie replied:
Just noticed there's a CAN linux params struct defined below. Those 2 should be kept in sync.

vincent-d replied:
`CAN_DLL_NUMOF` is defined by a `CFLAGS` parameter in `sys/Makefile.include`. The value is set from the `CAN_DLL_NUMOF` Makefile variable. There is no limitation with the number of interfaces (except the memory, obviously ;).

Then the `candev_linux_params.h` contains the RIOT interface configuration, the name in that file is the RIOT name for the interface, which is registered when the device thread is started (see `sys/can/device.c`)

Then we need a way to map from a RIOT dev to a linux dev. That's what the `candev_linux_conf[]` is made for. But I also have been requested to make that setting a command line parameter. It ended up with a hacky thing, I believe.
In my opinion the `candev_linux_conf[]` does not belong to that file but to have an architecture closer to the regular device driver, it should go the the auto_init file. But that array should be accessible from the `startup.c` file...

So I agree something is wrong here, but I don't have a cleaner solution to propose right now.

aabadie commented on cpu/native/can/candev_linux.c:
```diff
@@ -0,0 +1,550 @@
+static void _isr(candev_t *candev);
+static int _set(candev_t *candev, canopt_t opt, void *value, size_t value_len);
+static int _get(candev_t *candev, canopt_t opt, void *value, size_t max_len);
+static int _abort(candev_t *candev, const struct can_frame *frame);
```
Looking at the implementation of this function below, I don't understand the utility. Same for `_power_up` and `_power_down`. is it here for compatibility with real can devices ?

vincent-d replied:
Yes, `abort`, `power_up` and `power_down` make sense only for real CAN devices. The native candev is a bit simplified.

aabadie commented on cpu/native/include/candev_linux.h:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Maximum size of an interface name
+ */
+#define CAN_MAX_SIZE_INTERFACE_NAME (5)
```
This doesn't seem to be really useful since only 2 interfaces can be defined at most.

vincent-d replied:
This is the string length for storing the interface name, not the number of interfaces.

aabadie commented on drivers/trx/Makefile:
```diff
@@ -0,0 +1,3 @@
+MODULE = trx
```
This line is not needed

aabadie commented:
@vincent-d there are still issues regarding memory with nucleo32-l031 and nucleo-l053. @kaspar030 if we want to merge this, we might need to update the CI since the packet `libsocketcan-dev` is required for native.

vincent-d commented:
- Fixed `BOARD_INSUFFICIENT_MEMORY`
- Improved `conn_can`, close can be called from another thread
- Improve and simplify test app

aabadie commented on sys/include/can/isotp.h:
```diff
@@ -0,0 +1,197 @@
+    struct tpcon rx;               /**< receive state */
+    xtimer_t tx_timer;             /**< timer for tx operations */
+    xtimer_t rx_timer;             /**< timer for rx operations */
+    can_reg_entry_t entry;         /**< entry containig ifnum and upper layer msg system */
```
containing

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+/**
+ * @brief Allocate a CAN packet to transmit
+ *
+ * This function allocates a CAN packet and associate it to the @p ifnum and @p tx_pid.
```
associate**s**

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+#include "mbox.h"
+#endif
+
+typedef struct can_pkt can_pkt_t;
```
This line is not needed.
Your can just use :
```c
typedef struct {
    [attributes]
} can_packet_t;
```
(I don't think can_pkt is used elsewhere, didn't check)

aabadie commented on sys/include/can/pkt.h:
```diff
@@ -0,0 +1,164 @@
+ *
+ * This is used to allocate a return value to the upper layer
+ *
+ * @param data data which will be returned
```
[in]/[out] missing here and below

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,268 @@
+ *
+ *
+ * @file
+ * @brief       Definitions high-level CAN interface
```
Definitions **of**, CAN => RAW CAN

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,268 @@
+/**
+ * @brief Abort a CAN frame
+ *
+ * Abort the frame identifiedby @p handle in the interface @p ifnum
```
missing space between identified and by

aabadie commented on sys/include/can/raw.h:
```diff
@@ -0,0 +1,268 @@
+ *
+ * Abort the frame identifiedby @p handle in the interface @p ifnum
+ * If no tx confirmation is received, this function must be called by the upper layer
+ * to ensure the driver frees its tx mailbox. THe driver is not responsible of tx timeouts.
```
The

aabadie commented on sys/include/can/router.h:
```diff
@@ -0,0 +1,116 @@
+/**
+ * @brief Dispatch a RX indication to subscribers threads
+ *
+ * This function goes through the list of subsribed filters to send a message to each
```
subscribed

aabadie commented on tests/conn_can/README.md:
```diff
@@ -0,0 +1,146 @@
+Linux CAN basic commands
+========================
+
+Once the interfaces are set up, can-utils commands provide a way to send an receive
```
send and receive

aabadie commented on tests/conn_can/main.c:
```diff
@@ -0,0 +1,647 @@
+static int _send(int argc, char **argv)
+{
+    if (argc < 5) {
+        puts("Send: Not enough arguments");
```
maybe print command usage instead (here and below)

vincent-d commented:
- Fixed doxygen typos
- Fixed NULL pointers checks
- Clean up test app (print usage)

aabadie commented:
Thanks for adding the changes that fast @vincent-d. I tested again the sample application in the tests but the commands provided are not very intuitive (I did that in brainless user mode as usual ;) ). I could get all commands to work in the end, except the ISO-TP. 

Here is the error I had:
```
> isotprecv -s 708 -d 700 vcan0
socket: Protocol not supported
```

Do I need something special (like [this](https://github.com/hartkopp/can-isotp)) ?

vincent-d commented:
> Do I need something special (like this) ?

Yes, for ISO-TP, you need an experimental module as I wrote in the README ;)

You can test between two instances of RIOT without any additional module, though.

aabadie commented:
> Yes, for ISO-TP, you need an experimental module as I wrote in the README ;)

Ah yes, I missed that.

> You can test between two instances of RIOT without any additional module, though.

I did another try between 2 RIOT native instances but it didn't work:
* on instance 1:
```
> can bind_isotp 0 1 700 708
> can recv_isotp 1 10000000
```
* on instance 2:
```
> can bind_isotp 0 1 700 708
> can send_isotp 1 01 02 03 04 05 0a 0b 0c
```
But nothing arrives...
I always get "Error when sending"

vincent-d commented:
you need to swap the source and dest ids:
- on instance 1:
```
> can bind_isotp 0 1 700 708
> can recv_isotp 1 10000000
```
- on instance 2:
```
> can bind_isotp 0 1 708 700
> can send_isotp 1 01 02 03 04 05 0a 0b 0c
```

You can 'spy' the bus from Linux to see the traffic and check the frames are sent correctly:
```
candump vcan0
```

aabadie commented:
@vincent-d, thanks for the clarification, it works much better ! (and proves my poor knowledge of the CAN bus).

aabadie commented:
I think we are very close to get this merged. Squashing still remains and maybe other reviewers still have comments, @OlegHahm, @haukepetersen, @kaspar030, @PeterKietzmann what do you think ?

vincent-d commented:
While I was at it, I added basic send and dump shell commands. Though it can come in a follow up PR if you prefer.

I also cleaned up the `raw_can_get_*` function names and added a `raw_can_get_dev_by_ifnum()` to retrieve a candev descriptor pointer.

vincent-d commented:
I've open a PR to add libsocketCAN in our Docker image (https://github.com/RIOT-OS/riotdocker/pull/31). Hope it will allow the CI to build.

vincent-d commented:
@aabadie is that OK if I squash ? The commit list is really long now.

aabadie commented:
> is that OK if I squash ?

yes go ahead. By the way, this branch has conflicts and needs rebase

vincent-d commented:
Rebased and squashed!

vincent-d commented:
Ping @aabadie, @OlegHahm, @haukepetersen, @kaspar030, @PeterKietzmann

aabadie commented:
we can slso ping @miri64 ;)

kaspar030 commented:
plus 7k lines! that'll take while. @vincent-d thanks for your patience...

haukepetersen commented on sys/include/can/can.h:
```diff
@@ -0,0 +1,149 @@
+/**
+ * @brief Controller Area Network Identifier structure
+ *
+ * bit 0-28  : CAN identifier (11/29 bit)
```
a comment if the 11-bit ID is left or right aligned in those 29 bits would be useful...

vincent-d replied:
Addressed

miri64 commented:
> we can also ping @miri64 ;)

I don't think I'm comfortable reviewing this.

haukepetersen commented:
Hej, sorry for the slow review. To the current state of the PR if have only some minor remarks (see above). 

I general I have however the strong feeling, that this PR is extremely complex (IMHO much much more than it could be), so I have to admit that I understand only about 1/8 of the code, even with a pretty ok knowledge of CAN. IMHO in this case it does not seem like a good idea, to mimic GNRC and to build heavily on messaging and mboxes. But then again, I don't understand everything that this PR does...

But as the code does work, I won't be blocking it - I'd say lets merge it anyway and see in which direction this will develop in the future. 

I would only ask to open a tracking issue for the following items/todos:
- add more comprehensive documentation, mostly on explaining the high-level modules used in the stack and how they plug into each other
- migrate the `conn_can` interfaces to `sock_can` (consistent with `sock_xxx`)?!
- give the whole CAN stack a name (like done for GNRC) and split the stack specific implementations/internal interfaces from the generic CAN interfaces to allow for alternative CAN stack(s). 

So you have my ACK once the minor comments above are addressed.

vincent-d commented:
> IMHO in this case it does not seem like a good idea, to mimic GNRC and to build heavily on messaging and mboxes.

Could you elaborate a bit?

> So you have my ACK once the minor comments above are addressed.

Thanks! I already fixed the easy ones in the Makefiles, which must be rebase issues I assume.

haukepetersen commented:
> Could you elaborate a bit?

I try: as said, I have not understood everything in this PR fully, so I might misunderstand how things work here and there. But in general, it seems to me, that the CAN stack makes heavy use of the messaging and inter-thread communication. I know we do this in GNRC also, but it turned out that this is not the best approach. So if possible, I would try to keep the IPC stuff down to a minimum and rather build upon synchronous functions (that return their result instead of asynchronously waiting for the result through IPC). I think it would also be possible to build the CAN stack into a single thread, even when working with multiple CAN interfaces. But as said: this would be my design approach on this, this does not mean that the approach in this PR is not valid...

vincent-d commented:
Understood. And I agree on some parts. 

After some months hacking this, IPC sometimes gave me some headaches! But the design was inspired by gnrc, and I thought at the very beginning that we could reuse more code from gnrc. But it turned out that was too different.

Anyway, this approach works so far, and with a bit more of documentation, it would be a good starting point for CAN support. Hopefully, people will start using it, and maybe they could like ;)

I'll work on your remarks next week.

vincent-d commented:
Renamed trx to can_trx

vincent-d commented:
`power_up` and `power_down` functions moved to `set`

vincent-d commented on drivers/include/can/candev.h:
```diff
@@ -0,0 +1,203 @@
+     * @return              0 on success
+     * @return              <0 on error
+     */
+    int (*power_up)(candev_t *dev);
```
addressed

haukepetersen replied:
but is it my diff view or do the `power_xx` functions still exist in this interface?!

vincent-d commented:
I think all comments have been addressed so far.

vincent-d commented:
I gave a try to Murdock, which unsurprisingly complains about socketCAN being unavailable. Otherwise, it seems to pass.

cgundogan commented:
seems like `libsocketcan` is not installed on the murdock slaves. @smlng @kaspar030 ? https://github.com/RIOT-OS/RIOT/pull/5793/files#diff-5a358c10b7aef66f47e495cccf7908de

kaspar030 commented:
> seems like libsocketcan is not installed on the murdock slaves. @smlng @kaspar030 ?

https://github.com/RIOT-OS/riotdocker/pull/31 needs review. once merged, I'll update the workers.

emmanuelsearch commented:
@vincent-d what's the status on this?

vincent-d commented:
> @vincent-d what's the status on this?

It's ready as it is. Let's do as @haukepetersen proposed and open some issue to track the remaining points.

I can squash and and let's have RIOT-OS/riotdocker#31 merged for final Murdock approval!

vincent-d commented:
squashed and rebased

aabadie commented:
Needs squashing label was still set. I removed it and restarted Murdock. I guess I'll merge when everything is green (and after lunch :) )

vincent-d commented:
> I guess I'll merge when everything is green

We need libsocketcan to build on native,  RIOT-OS/riotdocker#31 is ready but needs review.

aabadie commented:
Ah yes, I forgot this...

aabadie commented:
@smlng @kaspar030 can you merge RIOT-OS/riotdocker#31 or can I do it myself ? Does the CI workers need special update afterwards or is it automatic once merged ?

emmanuelsearch commented:
CI still [complains](https://ci.riot-os.org/RIOT-OS/RIOT/5793/8ad59e35cf8fc38be954c4b4ac92b02ce53a4d88/output/compile/tests/conn_can/native.txt).

kaspar030 commented:
CI docker container is rebuilding as we speak.

vincent-d commented:
All green, finally! Thanks @kaspar030

aabadie commented:
Let's merge this one then ! GO!

aabadie commented:
Thanks for your patience @vincent-d :)

aabadie commented:
Now it's time to work on #6178, #6927 and #6925 for CAN hardware support. Those PRs need to be rebased.

vincent-d commented:
Thanks all!

> Now it's time to work on #6178, #6927 and #6925 for CAN hardware support.

And #6276 ;)

emmanuelsearch commented:
@vincent-d congrats!
